### ISSUE TITLE
meson: translated man pages

### ISSUE THREAD

**Issue by eworm-de** [util-linux/util-linux#2842]
Is there a way to generate and install translated man pages with meson? I could not find one.

**Comment by t-8ch** [comment]
I'm fairly sure right now this isn't implemented.

**Comment by eworm-de** [comment]
Then take this as a feature request... I have a downstream issue to provide translated man pages, currently I can not as we switched to meson already.

**Comment by jwillikers** [comment]
@t-8ch I've started poking at how to go about this, and it seems there's probably two different approaches to take: the deterministic Meson way and the just do everything in a script way.

The deterministic approach requires encoding which translations each man page is available for, which would look like how the `manlinks` variable is used.
Here's what it would look like for the `chfn` man page.

```
mantranslations += {'chfn': ['de', 'fr', 'sr', 'uk']}
```

Because translations are generated globally, this method would be more involved when updating translations. For instance, if the `cs` translations are updated adequately for `chfn` to be translated, then it will be necessary to add `cs` to this list. Of course, updating translations could effect multiple man pages, and the translations would then need to added for all of them. The good part about this approach is that things would break if we all of a sudden stopped producing a translation for a man page. There's also the benefit that this approach should be pretty straightforward to implement in the existing man page generation logic. This also means that it will also be very easy to control which translations are installed based on which tools are being built.

The alternative is to pretty much encode the majority of the translation work in a separate Python or shell script, since in Meson it would be necessary to hard-code the generated output files in order to create dependencies between custom targets. This method would require less bookkeeping, and be more complicated to implement. It would also require communicating which tools are enabled and disabled to the script to ensure only translations for tools that are actually built are installed.

I'm not familiar with `po4a`, so perhaps there's some alternative way of invoking the it that would work better with Meson. I've just been looking at using it with the config file as is done in the Makefile.

**Comment by t-8ch** [comment]
Thanks for looking into this.

Personally I'm not a big fan of hardcoding everything, this will need some process to update after the translations come in.
If the logic gets put into a script, that could be used from the autotools build, too, right?
One additional possibility would be to run based on po4a at meson configuration time and generate proper targets from that.

Unfortunately no possibility seems really great...

**Comment by karelzak** [comment]
There is already a hardcoded list of languages in po-man/po4a.cfg (maybe you can read `[po4a_langs]` line from meson) . 

BTW, I found that two languages needed to be added to the list. Hardcoding anything is bad ...

**Comment by jwillikers** [comment]
> There is already a hardcoded list of languages in po-man/po4a.cfg (maybe you can read `[po4a_langs]` line from meson) .
> 
> BTW, I found that two languages needed to be added to the list. Hardcoding anything is bad ...

I can parse those out at configure time, so that shouldn't be a problem.

> Personally I'm not a big fan of hardcoding everything, this will need some process to update after the translations come in. If the logic gets put into a script, that could be used from the autotools build, too, right?

I think so.

> One additional possibility would be to run based on po4a at meson configuration time and generate proper targets from that.

The configure time approach would work, I think, but wouldn't cope very well if someone is working on translations and more or less files are created than were at configure time.
I found a flag, `--translate-only`, for `po4a` that allows you to generate the translations for an individual file.
I'm starting to think that creating a target per locale and adoc file using this command would work best.
To get this to work, something needs to handle when the file isn't generated due to an insufficient translation percentage.
A small wrapper script can probably just generate an empty or minimal file in that case, unless I find a way for Meson to somehow handle not generating a file for a custom target.
Asciidoctor doesn't like being asked to generate a man page from an empty file, so a another wrapper or a minimum viable adoc file might be necessary to accommodate it.
This just leaves the last hurdle, which is to install only the non-empty / non-minimal man pages.
A script that runs at install time seems the most likely solution.

This approach handles generating translations for only the tools enabled as part of the build and it should be possible to track dependencies properly among everything to properly trigger rebuilding translations when adoc or po files are modified.
Nothing needs be hard-coded either.

Thoughts on that?

**Comment by t-8ch** [comment]
> The configure time approach would work, I think, but wouldn't cope very well if someone is working on translations and more or less files are created than were at configure time.

I don't think the translators work like that.
@karelzak uploads the translation files to translationproject.org, the translators work there and @karelzak reimports the updated translation files into the source tree.
If there are hardcoded configurations then he will have to figure out after the reimport on how to adapt the configuration.

**Comment by jwillikers** [comment]
> > The configure time approach would work, I think, but wouldn't cope very well if someone is working on translations and more or less files are created than were at configure time.
> 
> I don't think the translators work like that. @karelzak uploads the translation files to translationproject.org, the translators work there and @karelzak reimports the updated translation files into the source tree. If there are hardcoded configurations then he will have to figure out after the reimport on how to adapt the configuration.

Oh, that's a much simpler situation then. I was thinking that perhaps they were developed iteratively as part of the codebase.

**Comment by jwillikers** [comment]
Of course, since translations are only valid if they reach 80%, changes to individual man pages or the common man files could effect whether a translated adoc file is generated or not.

**Comment by jwillikers** [comment]
I hit https://github.com/mesonbuild/meson/issues/2320 trying to get per-file targets working...
That seems to mean the script route it is.

I've got something working for that now, but I can't adequately convey dependencies between the custom target that runs the script and the man adoc files.
Using `files()` for the `manadocs` just like is done for the source files makes the dependencies on the adoc files work.
That will take me a bit to convert things to use `files()` calls in each of the subdirectories for the man adoc files.
I've made a draft PR #3296 without the dependencies on the individual adoc files yet.
